examples/src/aleph/examples/http2.clj

```diff
@@ -0,0 +1,108 @@
+(ns aleph.examples.http2
+  (:require
+    [aleph.http :as http]
+    [aleph.netty :as netty]
+    [clj-commons.byte-streams :as bs])
+  (:import
+    (java.util.zip
+      GZIPInputStream
+      InflaterInputStream)))
+
+;; This file assumes you've already covered basic HTTP client/server usage
+;; in aleph.examples.http. This file covers just HTTP/2 additions.
+
+;; ## Servers
+
+;; By default, you don't usually start up an HTTP/2-only server. In most
+;; scenarios, the server will support multiple HTTP versions, and you don't
+;; know which you need until the client connects. Instead, you
+;; start up a general HTTP server with support for Application-Layer Protocol
+;; Negotiation (ALPN), which negotiates the HTTP version during the TLS
+;; handshake. Unlike HTTP/1, HTTP/2 effectively requires TLS.
+
+;; For Aleph, this means you need to provide a `SslContext` to the server
+;; with ALPN configured, and HTTP/2 available as a protocol.
+
+(defn hello-world-handler
+  "A basic Ring handler which immediately returns 'hello world!'"
+  [_]
+  {:status  200
+   :headers {"content-type" "text/plain"}
+   :body    "hello world!"})
+
+(def http2-ssl-ctx
+  "This creates a self-signed certificate SslContext that supports both HTTP/2
+   and HTTP/1.1. The protocol array lists which protocols are acceptable, and
+   the order indicates the preference; in this example, HTTP/2 is preferred over
+   HTTP/1 if the client supports both.
+
+   NB: Avoid self-signed certs unless you control the clients, too. Browsers
+   will not accept them. See aleph.netty/ssl-server-context."
+  (netty/self-signed-ssl-context
+    "localhost"
+    {:application-protocol-config (netty/application-protocol-config [:http2 :http1])}))
+
+
+(def s (http/start-server hello-world-handler
+                          {:port        443
+                           :ssl-context http2-ssl-ctx}))
+
+;; ## Clients
+
+;; Like the server, the client must specify which HTTP versions are acceptable.
+;; Since connections are typically shared/reused between requests, the standard
+;; way to do this is to set {:connection-options {:http-versions [:http2 :http1]}}
+;; when setting up a connection-pool. At the moment, Aleph's default connection
+;; pool is still HTTP/1-only, so you must specify a custom connection pool. This
+;; will change in a future version of Aleph.
```

Comment:
I would like to. If we can support some HTTP1 leftovers like multipart encoding and CONNECT, we should be able to make H2 a completely transparent drop-in replacement. 

Want to split the difference and say "will probably change"?

Reply:
It can be kept that way if you are sure, but I don't think we will change to default to HTTP2 within the following year.
We'll need to be sure the support of HTTP2 is correctly supported on the Aleph HTTP2 client... and it won't have any side effects.